packages/react/src/time-picker/utils.ts

```diff
@@ -11,15 +11,43 @@ export function filterOptions<Option extends DefaultComboboxOption>(
 
 	return options.filter(function filterOption(option) {
 		const HHmm = option.value.split(':').join('');
+
+		const isMatchWithValue = HHmm.includes(sanitizedInputValue);
+		const isMatchWithLabel = option.label
+			.toLowerCase()
+			.includes(sanitizedInputValue);
+		const isLessThan4Char = sanitizedInputValue.length < 4;
+
+		const formatted24hrTime = formatTime(sanitizedInputValue, 'HH:mm');
+		const [inputHour] = formatted24hrTime.split(':');
+		const [optionHour] = option.value.split(':');
+
+		const isFormattedTimeMatch = formatted24hrTime === option.value;
+
+		function checkIsA12thHour(hour: string) {
+			if (optionHour === '00') {
+				return ['00', '24'].includes(hour);
+			}
+			return hour === '12';
+		}
+
+		const isAnyHourMatch = optionHour === inputHour;
+
+		const is12thHourOption = checkIsA12thHour(optionHour);
+		const is12thHourInput = inputValue === '12' || checkIsA12thHour(inputHour);
+		const isA12thHourMatch = is12thHourOption && is12thHourInput;
+
+		const isRoughHourMatch =
+			isLessThan4Char && (isAnyHourMatch || isA12thHourMatch);
+
+		const isPartialLabelMatch = option.label.includes(inputValue);
+
 		const hasMatch =
-			HHmm.includes(sanitizedInputValue) ||
-			option.label.toLowerCase().includes(sanitizedInputValue) ||
-			// When 12 and 24 is typed, show 00:XX options
-			(HHmm.startsWith('00') &&
-				sanitizedInputValue.length < 4 &&
-				(sanitizedInputValue.startsWith('12') ||
-					sanitizedInputValue.startsWith('24'))) ||
-			formatTime(sanitizedInputValue, 'HH:mm') === option.value;
+			isMatchWithValue ||
```

Comment:
The beauty of the original approach was that it short-circuited the code and only ran more expensive things when applicable. This current change does lots of work (including formatting the time) on every keystroke.

I wonder if we can reduce this effort somehow… I'll take a look